# Database Final Project SQL Part

## What we have done:
## We have finished the basic database and add the Twitter social media data into our database. Now we will do the following parts to finish our final project

## 1. Add another social media data into our database, we choose the Youtube social media.

In [67]:
import pandas as pd

In [6]:
import os

import google.oauth2.credentials

from googleapiclient.discovery import build
from googleapiclient.errors import HttpError
from google_auth_oauthlib.flow import InstalledAppFlow

# The CLIENT_SECRETS_FILE variable specifies the name of a file that contains
# the OAuth 2.0 information for this application, including its client_id and
# client_secret.
CLIENT_SECRETS_FILE = "client_secret_889051474355-h7p14b366keb8c9hbklmomsniki7j7fp.apps.googleusercontent.com.json"

# This OAuth 2.0 access scope allows for full read/write access to the
# authenticated user's account and requires requests to use an SSL connection.
SCOPES = ['https://www.googleapis.com/auth/youtube.force-ssl']
API_SERVICE_NAME = 'youtube'
API_VERSION = 'v3'

def get_authenticated_service():
  flow = InstalledAppFlow.from_client_secrets_file(CLIENT_SECRETS_FILE, SCOPES)
  credentials = flow.run_console()
  return build(API_SERVICE_NAME, API_VERSION, credentials = credentials)

def channels_list_by_username(service, **kwargs):
  results = service.channels().list(
    **kwargs
  ).execute()
  
  print('This channel\'s ID is %s. Its title is %s, and it has %s views.' %
       (results['items'][0]['id'],
        results['items'][0]['snippet']['title'],
        results['items'][0]['statistics']['viewCount']))

if __name__ == '__main__':
  # When running locally, disable OAuthlib's HTTPs verification. When
  # running in production *do not* leave this option enabled.
  os.environ['OAUTHLIB_INSECURE_TRANSPORT'] = '1'
  service = get_authenticated_service()
  channels_list_by_username(service,
      part='snippet,contentDetails,statistics',
      forUsername='GoogleDevelopers')

Please visit this URL to authorize this application: https://accounts.google.com/o/oauth2/auth?response_type=code&client_id=889051474355-h7p14b366keb8c9hbklmomsniki7j7fp.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fyoutube.force-ssl&state=7oJPiplXtedUZ2c5EKunn7EBYyyhoX&prompt=consent&access_type=offline
Enter the authorization code: 4/AABXmlNr6oHS5Aco9uW1PX6CFGdB_2Kos4MupN7NG_TRDjzzs6dG0bc
This channel's ID is UC_x5XG1OV2P6uZZ5FSM9Ttw. Its title is Google Developers, and it has 132019135 views.


In [8]:
#!/usr/bin/python

from apiclient.discovery import build
from apiclient.errors import HttpError
from oauth2client.tools import argparser


# Set DEVELOPER_KEY to the API key value from the APIs & auth > Registered apps
# tab of
#   https://cloud.google.com/console
# Please ensure that you have enabled the YouTube Data API for your project.
DEVELOPER_KEY = "AIzaSyAYn2FhemN1zX6ER8ES4LlbSZC1OoIj1jQ"
YOUTUBE_API_SERVICE_NAME = "youtube"
YOUTUBE_API_VERSION = "v3"

youtube = build(YOUTUBE_API_SERVICE_NAME, YOUTUBE_API_VERSION,developerKey=DEVELOPER_KEY)

  # Call the search.list method to retrieve results matching the specified
  # query term.

videos = []
channels = []
playlists = []
W1=['TheShapeofWater', 'BlackPanther', 'RedSparrow', 'ThreeBillboards', 'Avengers', 'CallMeByYour', 'Jumanji', 'Annihilation', 'TombRaider',
    'ChrisEvans', 'RobertDowneyJr', 'jenniferlawrence', 'channing tatum', 'georgeclooney', 'johnnydepp','JaredLeto', 'RyanReynolds', 'ryangosling', 'ScarlettJohansson',
    'QuentinTarantino', 'BillyWilder', 'AkiraKurosawa', 'CharlieChaplin', 'HayaoMiyazaki', 'AlfredHitchcock', 'StanleyKubrick', 'MartinScorsese', 'StevenSpielberg', 'ChristopherNolan',
    'Oscar', 'Hollywood']
  # Add each result to the appropriate list, and then display the lists of
  # matching videos, channels, and playlists.
for i in W1:
    search_response = youtube.search().list(q=i,part="id,snippet",maxResults=50).execute()
    for search_result in search_response.get("items", []):
        if search_result["id"]["kind"] == "youtube#video":
            a=youtube.videos().list(part='snippet,contentDetails,statistics',id=search_result["id"]["videoId"]).execute()
            if 'tags' in a['items'][0]['snippet']:
                pass
            else:
                a['items'][0]['snippet']['tags'] = []
            if 'viewCount' in a['items'][0]['statistics']:
                pass
            else:
                a['items'][0]['statistics']['viewCount'] = []
            if 'commentCount' in a['items'][0]['statistics']:
                pass
            else:
                a['items'][0]['statistics']['commentCount'] = []
            if 'dislikeCount' in a['items'][0]['statistics']:
                pass
            else:
                a['items'][0]['statistics']['dislikeCount'] = []
            if 'likeCount' in a['items'][0]['statistics']:
                pass
            else:
                a['items'][0]['statistics']['likeCount'] = []
            videos.append((search_result["snippet"]["title"],search_result["id"]["videoId"],
                           a['items'][0]['contentDetails']['dimension'],a['items'][0]['contentDetails']['duration'],
                           a['items'][0]['snippet']['channelTitle'],a['items'][0]['snippet']['description'],
                           a['items'][0]['snippet']['publishedAt'],a['items'][0]['snippet']['tags'],
                           a['items'][0]['statistics']['commentCount'],a['items'][0]['statistics']['dislikeCount'],
                           a['items'][0]['statistics']['favoriteCount'],a['items'][0]['statistics']['likeCount'],
                           a['items'][0]['statistics']['viewCount']))



In [45]:
pd.set_option('display.max_colwidth', -1)
Youtubetable = pd.DataFrame(videos,columns=['title','link','dimension','duration','channel title','descreption','published time', 'tags','comment count', 'dislike count','favorite count','like count', 'view count'])

In [46]:
Youtubetable

title  \
0     THE SHAPE OF WATER | Official Trailer | FOX Searchlight                                                
1     THE SHAPE OF WATER Trailer (2017)                                                                      
2     Everything Wrong With The Shape of Water                                                               
3     รีบรีวิว THE SHAPE OF WATER นิทานชิ้นเอกของ Guillermo Del Toro #JUSTดูIT                               
4     THE SHAPE OF WATER | Final Trailer | FOX Searchlight                                                   
5     The Shape of Water Episode 1 English Dub                                                               
6     THE SHAPE OF WATER (2017) Ending Explained + Analysis                                                  
7     Guillermo del Toro's THE SHAPE OF WATER | All Clips and Trailer Compilation                            
8     17 Things I Missed In The Shape of Water Trailer                                                       
9     La Forma Del Agua (The Shape Of Water) - Trailer Subtitulado Español Latino 2017 Guillermo Del Toro    
10    THE SHAPE OF WATER | Extended Preview | FOX Searchlight                                                
11    The Shape of Water / Kiss Scene                                                                        
12    THE SHAPE OF WATER | "The Tale Of Love And Loss" TV Commercial | FOX Searchlight                       
13    THE SHAPE OF WATER Trailer (2017) Guillermo del Toro                                                   
14    Bishop Barron Comments on “The Shape of Water”                                                         
15    THE SHAPE OF WATER Behind The Scenes Featurettes                                                       
16    The Shape of Water Monster Explained                                                                   
17    The Shape of Water - Movie Review                                                                      
18    THE SHAPE OF WATER | Red Band Trailer | FOX Searchlight                                                
19    THE SHAPE OF WATER | "Embrace" TV Commercial | FOX Searchlight                                         
20    The Shape of Water Movie Clip - Movie Theater (2017) | Movieclips Coming Soon                          
21    Guillermo del Toro on The Shape of Water                                                               
22    25 Facts About The Shape of Water                                                                      
23    The Shape Of Water Best scene compilation | Elisa and Amphibian Man | Oscar winning movie              
24    คุยตัวอย่างหนัง แอบรักสัตว์ทดลอง!! the shape of water REACTION                                         
25    The Shape of Water Explained: References & Homages                                                     
26    55 Things You Missed In The Shape of Water (2017)                                                      
27    Guillermo del Toro, Octavia Spencer, Sally Hawkins, Michael Shannon 'The Shape of Water' | TIFF 2017   
28    The Shape Of Water "Making of" Featurettes (2017)                                                      
29    THE SHAPE OF WATER Red Band Trailer (2017)                                                             
...                                          ...                                                             
1490  MORRIS CHESTNUT PAST STORIES, MARITAL ISSUES  AND HOLLYWOOD STRUGGLE                                   
1491  MC Hollywood - Rapidamente Treme o Bumbum (KondZilla)                                                  
1492  Ghost Who Got The Juice (Hollywood Hines) Official Music Video Filmed by Chop Spot Production          
1493  Hollywood Movies in Hindi Dubbed 2018 | Full Action HD Hindi Dubbed Movies | Online Full Movies |      
1494  Hollywood Movies in Hindi Dubbed 2018 | Full Action HD Hindi Dubbed Movies | Full Length Movies        
1495  Hollywood Full Action Adv

## As you can see above, we choose the same tags used in the Twitter social media to scrape relative data from Youtube, and obtain the similar table called "Youtubetable".

## 2. Now we create another three tables for tags:
### 1> Synonyms table
### 2> Mis spelled words table
### 3> Semantic Information table

### Synonyms table

In [110]:
import csv
import pandas as pd
import pymysql
import tweepy
import json
from dateutil import parser
import pandas as pd
import re
import pymysql
import sqlalchemy
pd.set_option('display.max_colwidth', -1)

In [111]:
def sql_to_df(sql_query):
    df=pd.read_sql(sql_query,db)
    return df

In [112]:
db = pymysql.connect(host = '127.0.0.1', port = 3306, user = 'root', db = 'assignment2',charset='utf8mb4')
cursor = db.cursor()

In [16]:
a = '''SELECT tags FROM tags;'''
tags = sql_to_df(a)

In [17]:
tag = []
for i in range(len(tags)):
    tag.append(tags.iloc[i][0])

tag

['TheShapeofWater',
 'BlackPanther',
 'TheWalkingDead',
 'RedSparrow',
 'ThreeBillboards',
 'AvengersInfinityWar',
 'CallMeByYour',
 'Jumanji',
 'Annihilation',
 'TombRaider',
 'ChrisEvans',
 'RobertDowneyJr',
 'jenniferlawrence',
 'channing tatum',
 'georgeclooney',
 'johnnydepp',
 'JaredLeto',
 'RyanReynolds',
 'ryangosling',
 'ScarlettJohansson',
 'QuentinTarantino',
 'BillyWilder',
 'AkiraKurosawa',
 'CharlieChaplin',
 'HayaoMiyazaki',
 'AlfredHitchcock',
 'StanleyKubrick',
 'MartinScorsese',
 'StevenSpielberg',
 'ChristopherNolan',
 'Oscar',
 'Hollywood']

In [18]:
taglist = ['shape','panther','walker','sparrow','billboards',
           'avenge','call','jungle','annihilation','raider',
           'evans','robert','lawrence','tatum','george',
           'johnny','Jared','Reynolds','gosling','johansson',
           'Tarantino','Billy','Kurosawa','Chaplin','Miyazaki',
           'Hitchcock','Stanley','Scorsese','Spielberg','Christopher',
           'oscar','hollywood']

In [19]:
from nltk.corpus import wordnet as wn

In [20]:
for i in taglist:   
    print(wn.synsets(i))
    print('\n')

[Synset('shape.n.01'), Synset('shape.n.02'), Synset('human_body.n.01'), Synset('shape.n.04'), Synset('form.n.07'), Synset('condition.n.05'), Synset('supreme_headquarters_allied_powers_europe.n.01'), Synset('form.n.03'), Synset('determine.v.02'), Synset('shape.v.02'), Synset('shape.v.03')]


[Synset('jaguar.n.01'), Synset('panther.n.02'), Synset('cougar.n.01')]


[Synset('walker.n.01'), Synset('walker.n.02'), Synset('pedestrian.n.01'), Synset('walker.n.04'), Synset('walker.n.05'), Synset('walker.n.06')]


[Synset('sparrow.n.01'), Synset('hedge_sparrow.n.01')]


[Synset('billboard.n.01')]


[Synset('revenge.v.01')]


[Synset('call.n.01'), Synset('call.n.02'), Synset('cry.n.01'), Synset('call.n.04'), Synset('birdcall.n.01'), Synset('call.n.06'), Synset('margin_call.n.01'), Synset('call.n.08'), Synset('call.n.09'), Synset('call.n.10'), Synset('call.n.11'), Synset('call.n.12'), Synset('call_option.n.02'), Synset('name.v.01'), Synset('call.v.02'), Synset('call.v.03'), Synset('shout.v.02'), S

In [21]:
synonmys = [['human_body','form','condition','supreme_headquarters_allied_powers_europe','determine'],
            ['jaguar','panther','cougar','null','null'],
            ['pedestrian','walker','null','null','null'],
            ['sparrow','hedge_sparrow','null','null','null'],
            ['billboard','null','null','null','null'],
            ['revenge','null','null','null','null'],
            ['cry','birdcall','margin_call','call_option','shout'],
            ['jungle','hobo_camp','null','null','null'],
            ['annihilation','null','null','null','null'],
            ['plunderer','raider','null','null','null'],
            ['evans','null','null','null','null'],
            ['robert','null','null','null','null'],
            ['lawrence','null','null','null','null'],
            ['tatum','null','null','null','null'],
            ['george','null','null','null','null'],
            ['rebel','null','null','null','null'],
            ['clash', 'jolt','jar','null','null'],
            ['reynolds','null','null','null','null'],
            ['gosling','null','null','null','null'],
            ['null','null','null','null','null'],
            ['tarantino','null','null','null','null'],
            ['truncheon','billy','null','null','null'],
            ['kurosawa','null','null','null','null'],
            ['chaplin','null','null','null','null'],
            ['null','null','null','null','null'],
            ['hitchcook','null','null','null','null'],
            ['stanley','null','null','null','null'],
            ['scorsese','null','null','null','null'],
            ['spielberg','null','null','null','null'],
            ['christopher','null','null','null','null'],
            ['academy_award','null','null','null','null'],
            ['hollywood','null','null','null','null']
           ]

In [23]:
tagdf = pd.DataFrame(tag)

In [24]:
sy = pd.DataFrame(synonmys)

In [25]:
sy

0              1            2  \
0   human_body     form           condition     
1   jaguar         panther        cougar        
2   pedestrian     walker         null          
3   sparrow        hedge_sparrow  null          
4   billboard      null           null          
5   revenge        null           null          
6   cry            birdcall       margin_call   
7   jungle         hobo_camp      null          
8   annihilation   null           null          
9   plunderer      raider         null          
10  evans          null           null          
11  robert         null           null          
12  lawrence       null           null          
13  tatum          null           null          
14  george         null           null          
15  rebel          null           null          
16  clash          jolt           jar           
17  reynolds       null           null          
18  gosling        null           null          
19  null           null           null          
20  tarantino      null           null          
21  truncheon      billy          null          
22  kurosawa       null           null          
23  chaplin        null           null          
24  null           null           null          
25  hitchcook      null           null          
26  stanley        null           null          
27  scorsese       null           null          
28  spielberg      null           null          
29  christopher    null           null          
30  academy_award  null           null          
31  hollywood      null           null          

                                            3          4  
0   supreme_headquarters_allied_powers_europe  determine  
1   null                                       null       
2   null                                       null       
3   null                                       null       
4   null                                       null       
5   null                                       null       
6   call_option                                shout      
7   null                                       null       
8   null                                       null       
9   null                                       null       
10  null                                       null       
11  null                                       null       
12  null                                       null       
13  null                                       null       
14  null                                       null       
15  null                                       null       
16  null                                       null       
17  null                                       null       
18  null                                       null       
19  null                                       null       
20  null                                       null       
21  null                                       null       
22  null                                       null       
23  null                                       null       
24  null                                       null       
25  null                                       null       
26  null                                       null       
27  null                                       null       
28  null                                       null       
29  null                                       null       
30  null                                       null       
31  null                                       null

In [29]:
tag_synonmys_table = pd.concat( [tagdf, sy], axis=1 )

In [30]:
tag_synonmys_table.columns = ['tags','First synonmy', 'Second synonmy','Third synonmy','Fourth synonmy','Fifth synonmy']

In [31]:
tag_synonmys_table

tags  First synonmy Second synonmy Third synonmy  \
0   TheShapeofWater      human_body     form           condition      
1   BlackPanther         jaguar         panther        cougar         
2   TheWalkingDead       pedestrian     walker         null           
3   RedSparrow           sparrow        hedge_sparrow  null           
4   ThreeBillboards      billboard      null           null           
5   AvengersInfinityWar  revenge        null           null           
6   CallMeByYour         cry            birdcall       margin_call    
7   Jumanji              jungle         hobo_camp      null           
8   Annihilation         annihilation   null           null           
9   TombRaider           plunderer      raider         null           
10  ChrisEvans           evans          null           null           
11  RobertDowneyJr       robert         null           null           
12  jenniferlawrence     lawrence       null           null           
13  channing tatum       tatum          null           null           
14  georgeclooney        george         null           null           
15  johnnydepp           rebel          null           null           
16  JaredLeto            clash          jolt           jar            
17  RyanReynolds         reynolds       null           null           
18  ryangosling          gosling        null           null           
19  ScarlettJohansson    null           null           null           
20  QuentinTarantino     tarantino      null           null           
21  BillyWilder          truncheon      billy          null           
22  AkiraKurosawa        kurosawa       null           null           
23  CharlieChaplin       chaplin        null           null           
24  HayaoMiyazaki        null           null           null           
25  AlfredHitchcock      hitchcook      null           null           
26  StanleyKubrick       stanley        null           null           
27  MartinScorsese       scorsese       null           null           
28  StevenSpielberg      spielberg      null           null           
29  ChristopherNolan     christopher    null           null           
30  Oscar                academy_award  null           null           
31  Hollywood            hollywood      null           null           

                               Fourth synonmy Fifth synonmy  
0   supreme_headquarters_allied_powers_europe  determine     
1   null                                       null          
2   null                                       null          
3   null                                       null          
4   null                                       null          
5   null                                       null          
6   call_option                                shout         
7   null                                       null          
8   null                                       null          
9   null                                       null          
10  null                                       null          
11  null                                       null          
12  null                                       null          
13  null                                       null          
14  null                                       null          
15  null                                       null          
16  null                                       null          
17  null                                       null          
18  null                                       null          
19  null                                       null          
20  null                                       null          
21  null                                       null          
22  null                                       null          
23  null                                       null          
24  null                                       null          
25  null                                       null          
2

### Mis spelled words table
#### We set the cutoff value as 0.8,and if the index of similarity is greater than 0.8, it should be concluded into mis-spelled table

In [33]:
from difflib import SequenceMatcher as sm

In [34]:
misspell = []
for x in tag:
    mis = []
    for i in range(len(x)):   
        if x[i] == 'a':
            if sm(a=x,b=x[:i]+'e'+x[i+1:]).ratio() > 0.8:
                mis.append(x[:i] + 'e' + x[i+1:])
        if x[i] == 'e':
            if sm(a=x,b=x[:i]+'a'+x[i+1:]).ratio() > 0.8:
                mis.append(x[:i] + 'a' + x[i+1:])
        if x[i] == 'c':
            if sm(a=x,b=x[:i]+'k'+x[i+1:]).ratio() > 0.8:
                mis.append(x[:i] + 'k' + x[i+1:])
        if x[i] == 'k':
            if sm(a=x,b=x[:i]+'c'+x[i+1:]).ratio() > 0.8:
                mis.append(x[:i] + 'c' + x[i+1:])
        if x[i] == 'e':
            if sm(a=x,b=x[:i]+'i'+x[i+1:]).ratio() > 0.8:
                mis.append(x[:i] + 'i' + x[i+1:])
        if x[i] == 'i':
            if sm(a=x,b=x[:i]+'e'+x[i+1:]).ratio() > 0.8:
                mis.append(x[:i] + 'e' + x[i+1:])

        if i <= len(x) - 2:
            if x[i+1] == x[i]:
                if sm(a=x,b=x[0:i]+x[i+1:]).ratio() > 0.8:
                    mis.append(x[0:i]+x[i+1:])

        if i == len(x) - 1 and x[i] == 'e':
            if sm(a=x,b=x+'r').ratio() > 0.8:
                mis.append(x+"r")
        if len(mis) >= 5:
            break
    misspell.append(mis)

In [103]:
mis_spelled_table = pd.DataFrame(misspell)
mis_spelled_table = pd.concat( [tagdf, mis_spelled_table], axis=1 )
mis_spelled_table.columns = ['tags','First mis spelled', 'Second mis spelled','Third mis spelled','Fourth mis spelled','Fifth mis spelled','Sixth mis spelled']
mis_spelled_table

tags    First mis spelled   Second mis spelled  \
0   TheShapeofWater      ThaShapeofWater      ThiShapeofWater       
1   BlackPanther         BleckPanther         BlakkPanther          
2   TheWalkingDead       ThaWalkingDead       ThiWalkingDead        
3   RedSparrow           RadSparrow           RidSparrow            
4   ThreeBillboards      ThraeBillboards      ThrieBillboards       
5   AvengersInfinityWar  AvangersInfinityWar  AvingersInfinityWar   
6   CallMeByYour         CellMeByYour         CalMeByYour           
7   Jumanji              Jumenji              Jumanje               
8   Annihilation         Anihilation          Annehilation          
9   TombRaider           TombReider           TombRaeder            
10  ChrisEvans           ChresEvans           ChrisEvens            
11  RobertDowneyJr       RobartDowneyJr       RobirtDowneyJr        
12  jenniferlawrence     janniferlawrence     jinniferlawrence      
13  channing tatum       khanning tatum       chenning tatum        
14  georgeclooney        gaorgeclooney        giorgeclooney         
15  johnnydepp           johnydepp            johnnydapp            
16  JaredLeto            JeredLeto            JaradLeto             
17  RyanReynolds         RyenReynolds         RyanRaynolds          
18  ryangosling          ryengosling          ryangosleng           
19  ScarlettJohansson    SkarlettJohansson    ScerlettJohansson     
20  QuentinTarantino     QuantinTarantino     QuintinTarantino      
21  BillyWilder          BellyWilder          BilyWilder            
22  AkiraKurosawa        AciraKurosawa        AkeraKurosawa         
23  CharlieChaplin       CherlieChaplin       CharleeChaplin        
24  HayaoMiyazaki        HeyaoMiyazaki        HayeoMiyazaki         
25  AlfredHitchcock      AlfradHitchcock      AlfridHitchcock       
26  StanleyKubrick       StenleyKubrick       StanlayKubrick        
27  MartinScorsese       MertinScorsese       MartenScorsese        
28  StevenSpielberg      StavenSpielberg      StivenSpielberg       
29  ChristopherNolan     ChrestopherNolan     ChristopharNolan      
30  Oscar                None                 None                  
31  Hollywood            Holywood             Hollywod              

      Third mis spelled   Fourth mis spelled    Fifth mis spelled  \
0   TheShepeofWater      TheShapaofWater      TheShapiofWater       
1   BlaccPanther         BlackPenther         BlackPanthar          
2   TheWelkingDead       TheWalcingDead       TheWalkengDead        
3   RedSperrow           RedSparow            None                  
4   ThreBillboards       ThreaBillboards      ThreiBillboards       
5   AvengarsInfinityWar  AvengirsInfinityWar  AvengersInfenityWar   
6   CallMaByYour         CallMiByYour         None                  
7   None                 None                 None                  
8   Annihelation         Annihiletion         Annihilateon          
9   TombRaidar           TombRaidir           None                  
10  None                 None                 None                  
11  RobertDownayJr       RobertDowniyJr       None                  
12  jeniferlawrence      jenneferlawrence     jennifarlawrence      
13  chaning tatum        channeng tatum       channing tetum        
14  georgaclooney        georgiclooney        georgeklooney         
15  johnnydipp           johnnydep            None                  
16  JaridLeto            JaredLato            JaredLito             
17  RyanRiynolds         None                 None                  
18  None                 None                 None                  
19  ScarlattJohansson    ScarlittJohansson    ScarletJohansson      
20  QuentenTarantino     QuentinTerantino     QuentinTarentino      
21  BillyWelder          BillyWildar          BillyWildir           
22  AkireKurosawa        AkiraKurosewa        AkiraKurosawe         
23  CharliaChaplin       CharliiChaplin       CharlieCheplin        
24  HayaoMeyazak

### Semantic Information table

In [36]:
title = '''SELECT * FROM title;'''
title=sql_to_df(title)

name = '''select * from names'''
name = sql_to_df(name)

In [37]:
tag = ['The Shape of Water',
 'Black Panther',
 'The Walking Dead',
 'Red Sparrow',
 'Three Billboards',
 'Avengers InfinityWar',
 'Call Me By Your Name',
 'Jumanji welcome to jungel',
 'Annihilation',
 'Tomb Raider',
 'Chris Evans',
 'Robert Downey',
 'jennifer lawrence',
 'channing tatum',
 'george clooney',
 'johnny depp',
 'Jared Leto',
 'Ryan Reynolds',
 'ryan gosling',
 'Scarlett Johansson',
 'Quentin Tarantino',
 'Billy Wilder',
 'Akira Kurosawa',
 'Charlie Chaplin',
 'Miyazaki',
 'Alfred',
 'Stanley Kubrick',
 'Martin Scorsese',
 'Steven Spielberg',
 'Christopher Nolan',
 'Oscar',
 'Hollywood']

In [39]:
meaning = []
for t in tag:
    
    p = False        
    for i in range(len(title)):
        c=title['originalTitle'][i]
        if sm(a=t,b=c).ratio() > 0.75:
            meaning.append(title['genres'][i])
            print(t, ":", title['genres'][i])
            p=True
            
        if p == True:
                break

    for j in range(len(name)):
        d = name['primaryName'][j]
        print
        if sm(a = t, b = d).ratio()>0.7:
            meaning.append(name['primaryProfession'][j])
            print(t, ":", name['primaryProfession'][j])
            p=True
            
        if p == True:
                break 

The Shape of Water : Documentary
Black Panther : Action,Adventure,Sci-Fi
The Walking Dead : Comedy
Red Sparrow : Thriller
Three Billboards : Documentary
Avengers InfinityWar : Action,Adventure,Fantasy
Call Me By Your Name : Drama,Romance
Jumanji welcome to jungel : Action,Adventure,Comedy
Annihilation : Adventure,Drama,Fantasy
Tomb Raider : Adventure
Chris Evans : actor
Robert Downey : actor,producer,soundtrack
jennifer lawrence : actress,producer,director
channing tatum : actor,producer,soundtrack
george clooney : actor,producer,director
johnny depp : actor,producer,soundtrack
Jared Leto : actor,soundtrack,producer
Ryan Reynolds : actor,producer,soundtrack
ryan gosling : actor,soundtrack,producer
Scarlett Johansson : actor
Quentin Tarantino : writer,actor,producer
Billy Wilder : actor,producer
Akira Kurosawa : actress
Charlie Chaplin : Comedy
Miyazaki : actress,soundtrack
Alfred : None
Stanley Kubrick : director,writer,producer
Martin Scorsese : producer,director,actor
Steven Spielber

In [104]:
semantic_information_table = pd.DataFrame(meaning)
semantic_information_table = pd.concat( [tagdf, semantic_information_table], axis=1 )
semantic_information_table.columns = ['tags', 'semantic information']
semantic_information_table

tags               semantic information
0   TheShapeofWater      Documentary                      
1   BlackPanther         Action,Adventure,Sci-Fi          
2   TheWalkingDead       Comedy                           
3   RedSparrow           Thriller                         
4   ThreeBillboards      Documentary                      
5   AvengersInfinityWar  Action,Adventure,Fantasy         
6   CallMeByYour         Drama,Romance                    
7   Jumanji              Action,Adventure,Comedy          
8   Annihilation         Adventure,Drama,Fantasy          
9   TombRaider           Adventure                        
10  ChrisEvans           actor                            
11  RobertDowneyJr       actor,producer,soundtrack        
12  jenniferlawrence     actress,producer,director        
13  channing tatum       actor,producer,soundtrack        
14  georgeclooney        actor,producer,director          
15  johnnydepp           actor,producer,soundtrack        
16  JaredLeto            actor,soundtrack,producer        
17  RyanReynolds         actor,producer,soundtrack        
18  ryangosling          actor,soundtrack,producer        
19  ScarlettJohansson    actor                            
20  QuentinTarantino     writer,actor,producer            
21  BillyWilder          actor,producer                   
22  AkiraKurosawa        actress                          
23  CharlieChaplin       Comedy                           
24  HayaoMiyazaki        actress,soundtrack               
25  AlfredHitchcock      None                             
26  StanleyKubrick       director,writer,producer         
27  MartinScorsese       producer,director,actor          
28  StevenSpielberg      producer,director,cinematographer
29  ChristopherNolan     Adventure,Animation,Comedy       
30  Oscar                Family                           
31  Hollywood            Horror

## 3. Insert these 4 tables into SQL database and show these four tables in Jupyter Notebook

In [41]:
yconnect = sqlalchemy.create_engine('mysql+pymysql://root:@127.0.0.1:3306/assignment2?charset=utf8')

In [69]:
for i in ['title','link','dimension','duration','channel title','descreption','published time', 'tags','comment count', 'dislike count','favorite count','like count', 'view count']:
    for j in Youtubetable[i]:
        if j == []:
            Youtubetable[i] = 1

In [105]:
pd.io.sql.to_sql(Youtubetable,'Youtubetable', yconnect, schema='assignment2')

/Users/baiyue/anaconda3/lib/python3.6/site-packages/pymysql/cursors.py:166: Warning: (1300, "Invalid utf8 character string: 'F09F93'")
  result = self._query(query)
/Users/baiyue/anaconda3/lib/python3.6/site-packages/pymysql/cursors.py:166: Warning: (1300, "Invalid utf8 character string: 'F09F98'")
  result = self._query(query)
/Users/baiyue/anaconda3/lib/python3.6/site-packages/pymysql/cursors.py:166: Warning: (1300, "Invalid utf8 character string: 'F09F8C'")
  result = self._query(query)
/Users/baiyue/anaconda3/lib/python3.6/site-packages/pymysql/cursors.py:166: Warning: (1300, "Invalid utf8 character string: 'F09F92'")
  result = self._query(query)
/Users/baiyue/anaconda3/lib/python3.6/site-packages/pymysql/cursors.py:166: Warning: (1300, "Invalid utf8 character string: 'F09F8D'")
  result = self._query(query)
/Users/baiyue/anaconda3/lib/python3.6/site-packages/pymysql/cursors.py:166: Warning: (1300, "Invalid utf8 character string: 'F09F94'")
  result = self._query(query)
/Users/bai

In [106]:
pd.io.sql.to_sql(tag_synonmys_table,'tag_synonmys_table', yconnect, schema='assignment2')

In [107]:
pd.io.sql.to_sql(mis_spelled_table,'mis_spelled_table', yconnect, schema='assignment2')

In [108]:
pd.io.sql.to_sql(semantic_information_table,'semantic_information_table', yconnect, schema='assignment2')

### Our 4 new tables:

### Youtubetable

In [113]:
a = '''SELECT * FROM Youtubetable;'''
sql_to_df(a)

index  \
0     0       
1     1       
2     2       
3     3       
4     4       
5     5       
6     6       
7     7       
8     8       
9     9       
10    10      
11    11      
12    12      
13    13      
14    14      
15    15      
16    16      
17    17      
18    18      
19    19      
20    20      
21    21      
22    22      
23    23      
24    24      
25    25      
26    26      
27    27      
28    28      
29    29      
...   ..      
1490  1490    
1491  1491    
1492  1492    
1493  1493    
1494  1494    
1495  1495    
1496  1496    
1497  1497    
1498  1498    
1499  1499    
1500  1500    
1501  1501    
1502  1502    
1503  1503    
1504  1504    
1505  1505    
1506  1506    
1507  1507    
1508  1508    
1509  1509    
1510  1510    
1511  1511    
1512  1512    
1513  1513    
1514  1514    
1515  1515    
1516  1516    
1517  1517    
1518  1518    
1519  1519    

                                                                                                     title  \
0     THE SHAPE OF WATER | Official Trailer | FOX Searchlight                                                
1     THE SHAPE OF WATER Trailer (2017)                                                                      
2     Everything Wrong With The Shape of Water                                                               
3     ???????? THE SHAPE OF WATER ??????????????? Guillermo Del Toro #JUST??IT                               
4     THE SHAPE OF WATER | Final Trailer | FOX Searchlight                                                   
5     The Shape of Water Episode 1 English Dub                                                               
6     THE SHAPE OF WATER (2017) Ending Explained + Analysis                                                  
7     Guillermo del Toro's THE SHAPE OF WATER | All Clips and Trailer Compilation                            
8     17 Things I Missed In The Shape of Water Trailer                                                       
9     La Forma Del Agua (The Shape Of Water) - Trailer Subtitulado Español Latino 2017 Guillermo Del Toro    
10    THE SHAPE OF WATER | Extended Preview | FOX Searchlight                                                
11    The Shape of Water / Kiss Scene                                                                        
12    THE SHAPE OF WATER | "The Tale Of Love And Loss" TV Commercial | FOX Searchlight                       
13    THE SHAPE OF WATER Trailer (2017) Guillermo del Toro                                                   
14    Bishop Barron Comments on “The Shape of Water”                                                         
15    THE SHAPE OF WATER Behind The Scenes Featurettes                                                       
16    The Shape of Water Monster Explained                                                                   
17    The Shape of Water - Movie Review                                                                      
18    THE SHAPE OF WATER | Red Band Trailer | FOX Searchlight                                                
19    THE SHAPE OF WATER | "Embrace" TV Commercial | FOX Searchlight                                         
20    The Shape of Water Movie Clip - Movie Theater (2017) | Movieclips Coming Soon                          
21    Guillermo del Toro on The Shape of Water                                                               
22    25 Facts About The Shape of Water                                                                      
23    The Shape Of Water Best scene compilation | Elisa and Amphibian Man | Oscar winning movie              
24    ??????????????? ????????????????!! the shape of water REACTION                                         
25    The Shape of Water Explained: References & Homages                                                     
26    55 Things You Missed In The Shape of Water (2017)                                                  

### Synonmys_table

In [114]:
a = '''SELECT * FROM tag_synonmys_table;'''
sql_to_df(a)

index                 tags First mis spelled Second mis spelled  \
0   0      TheShapeofWater      human_body        form                
1   1      BlackPanther         jaguar            panther             
2   2      TheWalkingDead       pedestrian        walker              
3   3      RedSparrow           sparrow           hedge_sparrow       
4   4      ThreeBillboards      billboard         null                
5   5      AvengersInfinityWar  revenge           null                
6   6      CallMeByYour         cry               birdcall            
7   7      Jumanji              jungle            hobo_camp           
8   8      Annihilation         annihilation      null                
9   9      TombRaider           plunderer         raider              
10  10     ChrisEvans           evans             null                
11  11     RobertDowneyJr       robert            null                
12  12     jenniferlawrence     lawrence          null                
13  13     channing tatum       tatum             null                
14  14     georgeclooney        george            null                
15  15     johnnydepp           rebel             null                
16  16     JaredLeto            clash             jolt                
17  17     RyanReynolds         reynolds          null                
18  18     ryangosling          gosling           null                
19  19     ScarlettJohansson    null              null                
20  20     QuentinTarantino     tarantino         null                
21  21     BillyWilder          truncheon         billy               
22  22     AkiraKurosawa        kurosawa          null                
23  23     CharlieChaplin       chaplin           null                
24  24     HayaoMiyazaki        null              null                
25  25     AlfredHitchcock      hitchcook         null                
26  26     StanleyKubrick       stanley           null                
27  27     MartinScorsese       scorsese          null                
28  28     StevenSpielberg      spielberg         null                
29  29     ChristopherNolan     christopher       null                
30  30     Oscar                academy_award     null                
31  31     Hollywood            hollywood         null                

   Third mis spelled                         Fourth mis spelled  \
0   condition         supreme_headquarters_allied_powers_europe   
1   cougar            null                                        
2   null              null                                        
3   null              null                                        
4   null              null                                        
5   null              null                                        
6   margin_call       call_option                                 
7   null              null                                        
8   null              null                                        
9   null              null                                        
10  null              null                                        
11  null              null                                        
12  null              null                                        
13  null              null                                        
14  null              null                                        
15  null              null                                        
16  jar               null                                        
17  null              null                                        
18  null              null                                        
19  null              null                                        
20  null              null                                        
21  null              null                                        
22  null              null                                        
23  null              null                          

### mis_spelled_table

In [115]:
a = '''SELECT * FROM mis_spelled_table;'''
sql_to_df(a)

index                 tags    First mis spelled   Second mis spelled  \
0   0      TheShapeofWater      ThaShapeofWater      ThiShapeofWater       
1   1      BlackPanther         BleckPanther         BlakkPanther          
2   2      TheWalkingDead       ThaWalkingDead       ThiWalkingDead        
3   3      RedSparrow           RadSparrow           RidSparrow            
4   4      ThreeBillboards      ThraeBillboards      ThrieBillboards       
5   5      AvengersInfinityWar  AvangersInfinityWar  AvingersInfinityWar   
6   6      CallMeByYour         CellMeByYour         CalMeByYour           
7   7      Jumanji              Jumenji              Jumanje               
8   8      Annihilation         Anihilation          Annehilation          
9   9      TombRaider           TombReider           TombRaeder            
10  10     ChrisEvans           ChresEvans           ChrisEvens            
11  11     RobertDowneyJr       RobartDowneyJr       RobirtDowneyJr        
12  12     jenniferlawrence     janniferlawrence     jinniferlawrence      
13  13     channing tatum       khanning tatum       chenning tatum        
14  14     georgeclooney        gaorgeclooney        giorgeclooney         
15  15     johnnydepp           johnydepp            johnnydapp            
16  16     JaredLeto            JeredLeto            JaradLeto             
17  17     RyanReynolds         RyenReynolds         RyanRaynolds          
18  18     ryangosling          ryengosling          ryangosleng           
19  19     ScarlettJohansson    SkarlettJohansson    ScerlettJohansson     
20  20     QuentinTarantino     QuantinTarantino     QuintinTarantino      
21  21     BillyWilder          BellyWilder          BilyWilder            
22  22     AkiraKurosawa        AciraKurosawa        AkeraKurosawa         
23  23     CharlieChaplin       CherlieChaplin       CharleeChaplin        
24  24     HayaoMiyazaki        HeyaoMiyazaki        HayeoMiyazaki         
25  25     AlfredHitchcock      AlfradHitchcock      AlfridHitchcock       
26  26     StanleyKubrick       StenleyKubrick       StanlayKubrick        
27  27     MartinScorsese       MertinScorsese       MartenScorsese        
28  28     StevenSpielberg      StavenSpielberg      StivenSpielberg       
29  29     ChristopherNolan     ChrestopherNolan     ChristopharNolan      
30  30     Oscar                None                 None                  
31  31     Hollywood            Holywood             Hollywod              

      Third mis spelled   Fourth mis spelled    Fifth mis spelled  \
0   TheShepeofWater      TheShapaofWater      TheShapiofWater       
1   BlaccPanther         BlackPenther         BlackPanthar          
2   TheWelkingDead       TheWalcingDead       TheWalkengDead        
3   RedSperrow           RedSparow            None                  
4   ThreBillboards       ThreaBillboards      ThreiBillboards       
5   AvengarsInfinityWar  AvengirsInfinityWar  AvengersInfenityWar   
6   CallMaByYour         CallMiByYour         None                  
7   None                 None                 None                  
8   Annihelation         Annihiletion         Annihilateon          
9   TombRaidar           TombRaidir           None                  
10  None                 None                 None                  
11  RobertDownayJr       RobertDowniyJr       None                  
12  jeniferlawrence      jenneferlawrence     jennifarlawrence      
13  chaning tatum        channeng tatum       channing tetum        
14  georgaclooney        georgiclooney        georgeklooney         
15  johnnydipp           johnnydep            None                  
16  JaridLeto            JaredLato            JaredLito             
17  RyanRiynolds         None                 None                  
18  None                 None                 None                  
19  ScarlattJohansson    ScarlittJohansson    ScarletJohansson      
20  QuentenTarantino     QuentinTerantino     

### semantic_information_table

In [117]:
a = '''SELECT * FROM semantic_information_table;'''
sql_to_df(a)

index                 tags               semantic information
0   0      TheShapeofWater      Documentary                      
1   1      BlackPanther         Action,Adventure,Sci-Fi          
2   2      TheWalkingDead       Comedy                           
3   3      RedSparrow           Thriller                         
4   4      ThreeBillboards      Documentary                      
5   5      AvengersInfinityWar  Action,Adventure,Fantasy         
6   6      CallMeByYour         Drama,Romance                    
7   7      Jumanji              Action,Adventure,Comedy          
8   8      Annihilation         Adventure,Drama,Fantasy          
9   9      TombRaider           Adventure                        
10  10     ChrisEvans           actor                            
11  11     RobertDowneyJr       actor,producer,soundtrack        
12  12     jenniferlawrence     actress,producer,director        
13  13     channing tatum       actor,producer,soundtrack        
14  14     georgeclooney        actor,producer,director          
15  15     johnnydepp           actor,producer,soundtrack        
16  16     JaredLeto            actor,soundtrack,producer        
17  17     RyanReynolds         actor,producer,soundtrack        
18  18     ryangosling          actor,soundtrack,producer        
19  19     ScarlettJohansson    actor                            
20  20     QuentinTarantino     writer,actor,producer            
21  21     BillyWilder          actor,producer                   
22  22     AkiraKurosawa        actress                          
23  23     CharlieChaplin       Comedy                           
24  24     HayaoMiyazaki        actress,soundtrack               
25  25     AlfredHitchcock      None                             
26  26     StanleyKubrick       director,writer,producer         
27  27     MartinScorsese       producer,director,actor          
28  28     StevenSpielberg      producer,director,cinematographer
29  29     ChristopherNolan     Adventure,Animation,Comedy       
30  30     Oscar                Family                           
31  31     Hollywood            Horror

## 4. Answer the questions for project:

In [121]:
def process(l):
    delnum = []
    for i in range(len(l)):
        if l[i] == '':
            delnum.append(i)
            continue
        if l[i] in l[:i]:
            delnum.append(i)
            continue
    for i in reversed(delnum):
        l.pop(i)
    return l 


## problem i

In [83]:
sql='''
select text, hashtags from tweettable where screen_name
'''
sql_to_df(sql)

text  \
0   RT @keithboykin: A TV game show host (Trump) married to a model (Melania) has an affair with a porn star (Stormy Daniels), hires a movie pr…        
1   RT @poetWOAgun: @2WizardofOz Their parents GREW-UP on VIOLENT HOLLYWOOD MOVIES that were tied to VIOLENT VIDEO GAMES and their children pla…        
2   RT @brkicks: adidas Dame 4 #BlackPanther customs by @mammothnamedjim https://t.co/hvuEP8SPdw                                                        
3   RT @ABC: "Black Panther" is less than $18 million shy of passing "The Avengers" as the highest-grossing superhero movie of all time. https:…        
4   The Walking Dead Season 8 - EPISODE 13 BLOODSHED at HILLTOP https://t.co/mVBGMLB5Od via @YouTubeGaming… https://t.co/RcW2dFbJz8                     
5   RT @Fern_Renville: @71KyleJordan @realDonaldTrump After all the stress he’s put us through? TOTALLY enjoying the start of his self-inflicte…        
6   #NowPlaying Frankie Goes To Hollywood - Relax on #011FM The Big 80s https://t.co/2tdbHp1uPB                                                         
7   RT @ajplus: With 35 million tweets, #BlackPanther is officially the most tweeted-about movie of all time. https://t.co/ua1Qo2ADv3                   
8   *NSYNC Announce Reunion Date For Hollywood Walk Of Fame Ceremony https://t.co/L3K0MqFMxP                                                            
9   RT @TheWalkingDead: Go behind the scenes of @OverkillsTWD gorgeous video game trailer:  https://t.co/vK7X4K0uN8\n\n(via @DenofGeekUS) https:/…      
10  RT @WhatUpJT: Thanos vs Avengers (2018)  https://t.co/fHfITkVOkM                                                                                    
11  RT @Avengers: Here's a look at six new @EmpireMagazine covers for Marvel Studios' "Avengers: #InfinityWar! Read more: https://t.co/GM2edXn0…        
12  RT @FerLlanos: Estado de Los Avengers tras los Kenjivideos @KenjiFujimoriH https://t.co/AUrJm5J3DP                                                  
13  RT @HeerJeet: Old Hollywood complaint: black movies are niche &amp; don't make enough money.\nNew Hollywood complaint: black movies make too muc…   

             hashtags  
0                      
1                      
2   BlackPanther       
3                      
4                      
5                      
6   NowPlaying 011FM   
7   BlackPanther       
8                      
9                      
10                     
11  InfinityWar        
12                     
13

### Problem i answer: the result is the hashtags that related to the user called ioTSC, which can be represent as what people are saying about him/her.

## problem ii

In [84]:
sql='''
select retweetcounts, favorites from tweettable where tweet_id = 976229529063600128
'''
sql_to_df(sql)

retweetcounts  favorites
0  623            943

### Problem ii answer: the result is the retweet count and favourite count of the tweet whose tweet_id is 976229529063600128 which is a tweet by ioTSC. By seeing the retweet count and favourite count, we can know how many people saw this post roughly which is a measure to the question that how viral is the post.

## Problem iii

In [85]:
sql='''
select followers, friends, favourites,retweetcounts, favorites from tweettable where tweet_id = 976229529063600128'''
sql_to_df(sql)

followers  friends  favourites  retweetcounts  favorites
0  814        3586     33218       623            943

### Problem iii answer: we also count the followers, friends, favourites of the user-ioTSC. The influence of a post can be measure by how many people can see the post, which include the people who didn't like or retweet the post. And the followers can give us a rough knowledge how influence is the post

## Problem iv

In [125]:
sql='''
select text,hashtags from tweettable where tweet_id=976229241414176768'''
sql_to_df(sql)

text  \
0  4 more weeks for @Avengers #infinitywar comes out in the uk and north america can't wait @marvel  and @Disney April 27th! #CaptainAmerica   

                      hashtags  
0  infinitywar CaptainAmerica

In [132]:
sql='''select text, hashtags from tweettable where hashtags like '%nfinity%'
'''
from difflib import SequenceMatcher
def similar(a, b):
    return SequenceMatcher(None, a, b).ratio()
r=sql_to_df(sql)
r
a=[]
for i,j in enumerate(r['hashtags']):
    if similar(j,'infinitywar')>0.75:
        a.append(r['text'][i])
pd.DataFrame(list(set(a)))

0
0   RT @chadwickboseman: New @Avengers: #InfinityWar trailer just dropped. See you April 27th ????? ? https://t.co/bAueXRZCYV https://t.co/BcuP…        
1   RT @Marvel: Here's a look at six new @EmpireMagazine covers for Marvel Studios' @Avengers: #InfinityWar! Read more: https://t.co/6zESjhsVhR…        
2   RT @Marvel: Watch the new Marvel Studios’ @Avengers: #InfinityWar trailer. Get your tickets now: https://t.co/UiS1rbqfAk https://t.co/QRwN1…        
3   RT @marvelmex: Avengers: #InfinityWar ha superado la venta de entradas anticipadas considerablemente. ¡Estos son los números!\nhttps://t.co/…       
4   #InfinityWar https://t.co/VGasw45j11                                                                                                                
5   RT @Avengers: Here's a look at six new @EmpireMagazine covers for Marvel Studios' "Avengers: #InfinityWar! Read more: https://t.co/GM2edXn0…        
6   Just came home from the cinema where the #InfinityWar trailer had been show ? omg goosebumps all over when the Aven… https://t.co/o11s3jLAIG        
7   RT @Avengers: Watch the new Marvel Studios’ "Avengers: #InfinityWar" trailer. Get your tickets now: https://t.co/kctg8VkHan https://t.co/bG…        
8   RT @TheWalkingDead: Marvel: "#InfinityWar is the most ambitious crossover event in history"\nUs: https://t.co/VnvIkao6yk                            
9   RT @tomhupdates: Tom Holland as Peter Parker at Empire's Avengers: Infinity War Covers. #InfinityWar https://t.co/Gt8GoGWNKq                        
10  RT @Dynamitejimi: The order to watch before #InfinityWar.\n\nCaptain America: TFA\nIron Man\nThe Incredible Hulk\nIron Man 2\nThor\nThe Avengers\nI…
11  RT @letitiawright: squad goals ?? x @Avengers \n#InfinityWar https://t.co/B8FCxrfGmk                                                                
12  RT @THR: New @Avengers: #InfinityWar trailer nabs third-biggest debut ever https://t.co/6j1e7PaZh3 https://t.co/wauXbe6XCS                          
13  RT @Fandango: Be the first to see @Avengers: #InfinityWar at an opening night fan event with exclusive footage and giveaways! ??https://t.c…        
14  When it looks like the Avengers will make their last stand at the end of #InfinityWar, but Thanos raises the infini… https://t.co/GzwJWvPJon

### Problem iv answer: we take the tweet whose tweet_id is 976229241414176768 as my post. the result is the posts that like my post.

## Problem v

In [88]:
sql='''
select screen_name from tweettable where hashtags like '%nfinity%'
'''
sql_to_df(sql)

screen_name
0    Jays1of1       
1    ayuaqilah      
2    Kaname_Lagann  
3    Gojira007X     
4    lurysedruol    
5    winterssoldear 
6    Kaname_Lagann  
7    heyz_12o       
8    kymccartney    
9    meelissaahhh_  
10   DariusBlu      
11   Lothiriel_1    
12   hussein66783   
13   GodsWeapon     
14   jimenez_315    
15   sweet_hs       
16   jimenez_315    
17   hussein66783   
18   PrincessChez_  
19   crdude45       
20   crdude45       
21   curlysatan     
22   GodsWeapon     
23   soyandresacosta
24   ofrancethemovie
25   JuanSanchezG   
26   kilinieee      
27   ofrancethemovie
28   RebelInTheDark 
29   Brenden_russo  
..             ...  
160  kitscarlett1994
161  MacSuzBax15    
162  MacSuzBax15    
163  BilaRito       
164  ktaegod        
165  Daisy_Buckenham
166  muqirfanrhzd   
167  victru02       
168  clifftothemound
169  KikeCard2      
170  Mrvl2yk        
171  3SwordstyleJB  
172  marvelousmags_ 
173  daredevil_eg   
174  VincentJinny1  
175  Bievixnia      
176  llewellyn_onyx 
177  aMieBeckz      
178  llewellyn_onyx 
179  Dat_boy_cyril  
180  daredevil_eg   
181  FlebileCarmen  
182  HopeSummer98   
183  jpc7992        
184  Smithryan8     
185  aboutnici      
186  komori_xiu     
187  mwapeimok      
188  Ruka_Freecss   
189  aboutnici      

[190 rows x 1 columns]

### Problem v answer: We choose the users who post tweets contain hashtags contain 'nifinity', so we can get hashtags related to 'Avengers: Infinity War'. And the result is the users who post similar posts that like mine.

## Problem vi

In [89]:
sql='''
select screen_name from tweettable where hashtags like '%nfinity%' and followers>10000
'''
sql_to_df(sql)

screen_name
0  milastinson
1  FootSniffah
2  ktaegod

### Problem vi answer: From the result in problem v, we get users whose followers is larger than 1000. Then the result is the users I should following.

## Problem vii

In [90]:
sql = '''SELECT id, tags, trend FROM assignment2.tags
                    where trend like "up";'''

sql_to_df(sql)

id            tags trend
0  tt2873282  RedSparrow      up  
1  nm0262635  ChrisEvans      up  
2  nm0000123  georgeclooney   up  
3  nm0005351  RyanReynolds    up  
4  nm0000040  StanleyKubrick  up  
5  04         Oscar           up

In [91]:
sql = '''SELECT id, tags, trend FROM assignment2.tags
                    where trend like "down";'''
sql_to_df(sql)

id                 tags trend
0   tt5580390  TheShapeofWater      down
1   tt1825683  BlackPanther         down
2   00         TheWalkingDead       down
3   tt5027774  ThreeBillboards      down
4   tt4154756  AvengersInfinityWar  down
5   tt5726616  CallMeByYour         down
6   tt2283362  Jumanji              down
7   tt2798920  Annihilation         down
8   tt1365519  TombRaider           down
9   nm0000375  RobertDowneyJr       down
10  nm2225369  jenniferlawrence     down
11  nm0000136  johnnydepp           down
12  nm0001467  JaredLeto            down
13  nm0331516  ryangosling          down
14  nm0000233  QuentinTarantino     down
15  01         CharlieChaplin       down
16  03         AlfredHitchcock      down
17  nm0000217  MartinScorsese       down
18  nm0000229  StevenSpielberg      down
19  05         Hollywood            down

In [92]:
sql = '''SELECT id, tags, trend FROM assignment2.tags
                    where trend like "no change";'''
sql_to_df(sql)

id               tags      trend
0  nm1475594  channing tatum     no change
1  nm0424060  ScarlettJohansson  no change
2  nm0000697  BillyWilder        no change
3  nm0000041  AkiraKurosawa      no change
4  02         HayaoMiyazaki      no change
5  nm0634240  ChristopherNolan   no change

### Problem vii answer: the result shows whether the tag is trending or not

## Problem viii

In [94]:
sql='''
select b.hashtags,count(b.hashtags) from title a, tweettable b 
where a.tconst=b.id and a.primaryTitle='Avengers: Infinity War' and hashtags!='' group by b.hashtags order by count(b.hashtags) desc limit 3
'''
sql_to_df(sql)

hashtags  count(b.hashtags)
0  InfinityWar            124              
1  Avengers InfinityWar   22               
2  AvengersInfinityWar    21

### Problem viii answer: we get top 3 most used hashtags that related to 'Avengers: Infinity War'. the result is the hashtags that I should add to my post.

## Problem ix

In [93]:
sql='''
select screen_name from tweettable where hashtags like '%nfinity%'
'''
sql_to_df(sql)

screen_name
0    Jays1of1       
1    ayuaqilah      
2    Kaname_Lagann  
3    Gojira007X     
4    lurysedruol    
5    winterssoldear 
6    Kaname_Lagann  
7    heyz_12o       
8    kymccartney    
9    meelissaahhh_  
10   DariusBlu      
11   Lothiriel_1    
12   hussein66783   
13   GodsWeapon     
14   jimenez_315    
15   sweet_hs       
16   jimenez_315    
17   hussein66783   
18   PrincessChez_  
19   crdude45       
20   crdude45       
21   curlysatan     
22   GodsWeapon     
23   soyandresacosta
24   ofrancethemovie
25   JuanSanchezG   
26   kilinieee      
27   ofrancethemovie
28   RebelInTheDark 
29   Brenden_russo  
..             ...  
160  kitscarlett1994
161  MacSuzBax15    
162  MacSuzBax15    
163  BilaRito       
164  ktaegod        
165  Daisy_Buckenham
166  muqirfanrhzd   
167  victru02       
168  clifftothemound
169  KikeCard2      
170  Mrvl2yk        
171  3SwordstyleJB  
172  marvelousmags_ 
173  daredevil_eg   
174  VincentJinny1  
175  Bievixnia      
176  llewellyn_onyx 
177  aMieBeckz      
178  llewellyn_onyx 
179  Dat_boy_cyril  
180  daredevil_eg   
181  FlebileCarmen  
182  HopeSummer98   
183  jpc7992        
184  Smithryan8     
185  aboutnici      
186  komori_xiu     
187  mwapeimok      
188  Ruka_Freecss   
189  aboutnici      

[190 rows x 1 columns]

### Problem ix answer: This question is similar to the question v. Thus, if someone in the result of problem vi follow me, I should follower him

## Problem x

In [95]:
sql='''
SELECT sec_to_time(AVG(time_to_sec(CAST(created_at as time)))) as time
from tweettable where created_at!='NaT' and followers>1000
'''
sql_to_df(sql)

/Users/baiyue/anaconda3/lib/python3.6/site-packages/pymysql/cursors.py:166: Warning: (1292, "Incorrect datetime value: 'NaT' for column 'created_at' at row 1")
  result = self._query(query)


time
0 22:50:56.036400

### Problem x answer: We calculated the average time of these tweets. And the result shows the best time to post. Because the time is UTC, to adjust it into EST, we should substract 5.

## Problem xi

In [96]:
sql='''
create view b as select text, count(text) as counts from tweettable group by text having count(text)>1 order by count(text) desc;
'''
sql_to_df(sql)

DatabaseError: Execution failed on sql '
create view b as select text, count(text) as counts from tweettable group by text having count(text)>1 order by count(text) desc;
': (1050, "Table 'b' already exists")

In [97]:
sql='''
create view a as select text, count(text) as counts from tweettable where text like '%https://%' group by text having count(text)>1 order by count(text) desc;
'''
sql_to_df(sql)

DatabaseError: Execution failed on sql '
create view a as select text, count(text) as counts from tweettable where text like '%https://%' group by text having count(text)>1 order by count(text) desc;
': (1050, "Table 'a' already exists")

In [98]:
sql_to_df('select count(a.text)/c from a,(select count(b.text) as c from b) c')

count(a.text)/c
0  0.5374

### Problem xi answer: because the result number is 0.5374149659863946, which is larger than 0.5, thus we should add url to my post.

## Problem xii

In [99]:
sql='''
select ((retweetcounts+favorites)/followers) from tweettable where tweet_id = 976229529063600128
'''
sql_to_df(sql)

((retweetcounts+favorites)/followers)
0  1.9238

### Problem xii answer: This result is calculated by the amount of people this post can infulence divided by the followers of the user. We get the number 1.9, which means that the post not only reach to the friends of the user but also the friends of his/her friends.